### PR TITLE
Add loongarch64 support

### PR DESCRIPTION
**Pull request by k0tran**
PR implementing loongarch64 support

Tested on loongarch64 rust 1.80.0\
Builds with debug and release flags\
Passes all the tests with debug and release flags

Notes to maintainer:
- loongarch32 is not supported by rust
- loongarch has 7-argument syscall. I was able to come up with some implementation but it's rather ugly in my opinion
- no `clobber_abi` because it clobbers more than we need

Might be useful:
- LoonArch64 ELF ABI https://loongson.github.io/LoongArch-Documentation/LoongArch-ELF-ABI-EN.html
- GLibc syscall implementation (Sysiphus repo) https://git.altlinux.org/gears/g/glibc.git?p=glibc.git;a=blob;f=sysdeps/unix/sysv/linux/loongarch/sysdep.h;h=e11cd865ce5d5960e637c0c00017bb9c38874595#l131

### CODE REVIEW

**Review by jasonwhite — CHANGES_REQUESTED** [review]
Awesome! I'm very skeptical that we actually need the 7 argument syscall at all. I did a little searching through the Linux kernel and didn't find a single syscall that takes 7 arguments (for any architecture). Am I missing something? I know that `man syscall(2)` says this architecture can have 7 arguments, but I don't see that actually happening in practice (in the Linux kernel). Is this a relic of old kernel versions?

I'd like to keep `SyscallArgs` at 6 arguments, since this is the common case. If I'm wrong about there not being a real life 7 argument syscall, then I would only expose it via a direct call to `syscall7`.

Also, it looks like `loongarch64-unknown-linux-gnu` is supported by [cross-rs](https://github.com/cross-rs/cross). Can you try to enable it in `.github/workflows/ci.yml`?

**Comment by jasonwhite** [comment]
NOTE: The CI failures for thumb mode are unrelated to this PR. It looks like a recent Rust version broke this.

**Comment by k0tran** [comment]
I found that for now there is no macro for defining 7-arg syscall
https://github.com/torvalds/linux/blob/1fb918967b56df3262ee984175816f0acb310501/include/linux/syscalls.h#L227

Will redo and force loongarch to not using syscall7

**Comment by jasonwhite** [comment]
Looks like the loongarch64 CI linker failure is an issue with cross-rs. I suspect this is just because the CI for this repo is using an outdated version of cross-rs as I see there have been some recent fixes for loongarch64.

```
syscalls-483deb5c79caa4ef.00d8fg4tdgpxcva2mmxyzpx89.rcgu.o: Relocations in generic ELF (EM: 258)
syscalls-483deb5c79caa4ef.00d8fg4tdgpxcva2mmxyzpx89.rcgu.o: error adding symbols: file in wrong format
```

If you don't want to solve this issue, then I'm okay with simply having it disabled for now.

**Comment by k0tran** [comment]
On my loongarch64 machine and cross 0.2.5 (on x86 with loongarch64 toolchain) compiles just fine. I'll look if I can update ci cross version

**Comment by k0tran** [comment]
After installing rust from rustup everything started failing as in CI. Cross version is the same (0.2.5). I think the main cause of the error is this warning: 
```
[cross] warning: `cross` does not provide a Docker image for target loongarch64-unknown-linux-gnu, specify a custom image in `Cross.toml`.
```

**Comment by k0tran** [comment]
Found the problem. Right now cross version from `cargo install cross` does not work, but master (`cargo install cross --git https://github.com/cross-rs/cross`) does, so CI should wait until next cross-rs release, I suppose.

File: src/macros.rs [hunk 1]
```diff
@@ -72,6 +72,21 @@ macro_rules! syscall {
...
+            $nr,
+            $a1 as usize,
+            $a2 as usize,
+            $a3 as usize,
+            $a4 as usize,
+            $a5 as usize,
+            $a6 as usize,
+            $a7 as usize,
+        )
+    };
```
Comment (jasonwhite) [inline src/macros.rs]: This is still referencing `syscall7`.

File: src/macros.rs [hunk 2]
```diff
@@ -143,4 +158,17 @@ macro_rules! raw_syscall {
...
+        $crate::raw::syscall7(
+            $nr as usize,
+            $a1 as usize,
+            $a2 as usize,
+            $a3 as usize,
+            $a4 as usize,
+            $a5 as usize,
+            $a6 as usize,
+        )
+    };
```
Comment (jasonwhite) [inline src/macros.rs]: Same here.

**Pull request merged**